src/modules/accounting.py

```diff
@@ -0,0 +1,251 @@
+import logging
+import time
+from collections import defaultdict
+from typing import Tuple, Union, List
+
+from hexbytes import HexBytes
+from web3 import Web3
+from web3.types import TxParams
+
+from src.contract_utils.frame import is_current_epoch_reportable, get_latest_reportable_epoch
+from src.contract_utils.lido_keys import get_lido_validators, get_lido_node_operators
+from src.contract_utils.withdrawal_queue import get_withdrawal_requests_wei_amount
+from src.contracts import contracts
+from src.metrics.prometheus.accounting import (
+    ACCOUNTING_ACTIVE_VALIDATORS,
+    ACCOUNTING_VALIDATORS_BALANCE,
+    ACCOUNTING_EXITED_VALIDATORS,
+    WC_BALANCE,
+    LAST_FINALIZED_WITHDRAWAL_REQUEST,
+    WEI_TO_RESERVE,
+)
+from src.modules.interface import OracleModule
+from src.providers.beacon import BeaconChainClient
+from src.web3_utils.tx_execution import check_transaction, sign_and_send_transaction
+
+from src.web3_utils.typings import SlotNumber, MergedLidoValidator
+
+from src.variables import GAS_LIMIT, ACCOUNT
+
+
+logger = logging.getLogger(__name__)
+
+
+class Accounting(OracleModule):
+    """Calculates total lido validators balance and reports it to execution layer"""
+
+    def __init__(self, web3: Web3, beacon_chain_client: BeaconChainClient):
+        logger.info({'msg': 'Initialize Oracle Accounting Module.'})
+
+        self._w3 = web3
+        self._beacon_chain_client = beacon_chain_client
+
+        self._update_beacon_specs()
+
+    def _update_beacon_specs(self, block_identifier: Union[str, HexBytes] = 'latest'):
+        (
+            self.epochs_per_frame,
+            self.slots_per_epoch,
+            self.seconds_per_slot,
+            self.genesis_time,
+        ) = contracts.oracle.functions.getBeaconSpec().call(block_identifier=block_identifier)
+        logging.info({
+            'msg': 'Update beacon specs.',
+            'epochs_per_frame': self.epochs_per_frame,
+            'slots_per_epoch': self.slots_per_epoch,
+            'seconds_per_slot': self.seconds_per_slot,
+            'genesis_time': self.genesis_time,
+        })
+
+    def run_module(self, slot: SlotNumber, block_hash: HexBytes):
+        """Check if epoch is reportable and try to send report if it is."""
+        self._update_beacon_specs(block_identifier=block_hash)
+
+        if not is_current_epoch_reportable(self._w3, contracts.oracle, slot, block_hash):
```

Comment:
This check in the two-phase oracle will be more difficult. We must consider that the oracle must be able to change its report. This assumes an emergency situation if we have a bug in the consensus algorithm. In this case we are supposed to change the consensus algorithm in the oracle daemon and ask the committee members to update. 

So at startup the oracle should check if the current report is finalized, and if not, it should rebuild the report. If the hashes do not match what it sent before, send a new